ports/geckolib/gecko_bindings/sugar/ownership.rs

```diff
@@ -0,0 +1,340 @@
+    }
+
+    #[inline]
+    /// Like Deref, but gives an Option
+    pub fn borrow_opt(self) -> Option<&'a T> {
+        if self.is_null() {
+            None
+        } else {
+            Some(unsafe { &*self.ptr })
+        }
+    }
+
+    #[inline]
+    /// Borrowed<GeckoType> -> Option<&Arc<ServoType>>
+    pub fn as_arc_opt<U>(&self) -> Option<&Arc<U>> where U: HasArcFFI<FFIType = T> {
+        unsafe {
+            if self.is_null() {
+                None
+            } else {
+                Some(transmute::<&Borrowed<_>, &Arc<_>>(self))
+            }
+        }
+    }
+
+    #[inline]
+    /// Converts a borrowed FFI reference to a borrowed Arc.
+    /// Panics on null.
+    ///
+    /// &Borrowed<GeckoType> -> &Arc<ServoType>
+    pub fn as_arc<U>(&self) -> &Arc<U> where U: HasArcFFI<FFIType = T> {
+        self.as_arc_opt().unwrap()
+    }
+
+    #[inline]
+    /// Borrowed<ServoType> -> Borrowed<GeckoType>
+    pub fn as_ffi(self) -> Borrowed<'a, <Self as HasFFI>::FFIType> where Self: HasSimpleFFI {
+        unsafe { transmute(self) }
+    }
+
+    #[inline]
+    /// Borrowed<GeckoType> -> Borrowed<ServoType>
+    pub fn from_ffi<U>(self) -> Borrowed<'a, U> where U: HasSimpleFFI<FFIType = T> {
+        unsafe { transmute(self) }
+    }
+
+    #[inline]
+    /// Borrowed<GeckoType> -> &ServoType
+    pub fn as_servo_ref<U>(self) -> Option<&'a U> where U: HasSimpleFFI<FFIType = T> {
+        self.borrow_opt().map(HasSimpleFFI::from_ffi)
+    }
+}
+
+impl<'a, T> BorrowedMut<'a, T> {
+    #[inline]
+    /// Like DerefMut, but gives an Option
+    pub fn borrow_mut_opt(self) -> Option<&'a mut T> {
+        // We have two choices for the signature here, it can either be
+        // Self -> Option<&'a mut T> or
+        // &'b mut Self -> Option<'b mut T>
+        // The former consumes the BorrowedMut (which isn't Copy),
+        // which can be annoying. The latter only temporarily
+        // borrows it, so the return value can't exit the scope
+        // even if Self has a longer lifetime ('a)
+        //
+        // This is basically the implicit "reborrow" pattern used with &mut
+        // not cleanly translating to our custom types.
+
+        // I've chosen the former solution -- you can manually convert back
+        // if you need to reuse the BorrowedMut.
+        if self.is_null() {
+            None
+        } else {
+            Some(unsafe { &mut *self.ptr })
+        }
+    }
+
+    #[inline]
+    /// BorrowedMut<GeckoType> -> &mut ServoType
+    pub fn as_servo_mut_ref<U>(self) -> Option<&'a mut U> where U: HasSimpleFFI<FFIType = T> {
+        self.borrow_mut_opt().map(HasSimpleFFI::from_ffi_mut)
+    }
+}
+
+// technically not
+impl<'a, T> Deref for BorrowedMut<'a, T> {
+    type Target = Borrowed<'a, T>;
+    fn deref(&self) -> &Self::Target {
+        unsafe { transmute(self) }
+    }
+}
+
+#[repr(C)]
+/// Gecko-FFI-safe Arc (T is an ArcInner).
+/// This can be null.
+/// Leaks on drop. Please don't drop this.
```

Comment:
Also, we don't have a rust compiler with MIR enabled yet, so maybe this can't be done right now, but maybe leaving a todo comment adding a `Drop` impl that asserts could be nice.